### ISSUE TITLE
Symfony Messenger Support

### ISSUE THREAD

**Issue by kbond** [kbond/schedule-bundle#7]
Currently, all due tasks run synchronously in the same process. Utilizing Symfony Messenger would allow for adding due tasks to a queue to run asynchronously. I believe this could partially solve #6.

I'm envisioning two available messenger modes:

1. `async`: simply send all due tasks to the message bus without regard to the order they are run.
2. `sync`: send due tasks to the queue but ensure they are run in the defined order.

Notes:

- Messenger would be opt-in and the "mode" defined globally
- `CompoundTask`'s should always be run synchronously regardless if you are using the `sync` or `async` strategy
- `RunTask` message for `async` mode - simply contains the task to run
- `RunTaskSync` message for `sync` mode and `CommandTask`'s - contains a "queue" of tasks in the order they should be run. Handler runs first command, pops the "queue" and dispatches the remainder, repeat until nothing left on the "queue"
- Possible "flow":
  - Initiator process:
    - Determine tasks to run (due or manual)
    - Dispatch `BeforeScheduleEvent` (will need to be aware that it is using a message bus to adjust logs/etc accordingly)
    - Run `filterSchedule` and `beforeSchedule` extensions
    - Loop through tasks to run:
      - Run `filterTask` extension to see if task should be run
      - Dispatch to messenger
  - Worker process:
    - Dispatch `BeforeTaskEvent`
    - Run `beforeTask` extension
    - Run task
    - Run `afterTask`/`onTaskSuccess`/`onTaskFailure` extensions
    - Dispatch `AfterTaskEvent`
    - Run `afterSchedule`/`onScheduleSuccess`/`onScheduleFailure` extensions
    - Dispatch `AfterScheduleEvent`

**Comment by PhilETaylor** [comment]
One thing to note here is that the server that runs `schedule:run` might not be the server/application that runs the actual job when its pulled from the messenger queue.

The commands available to Symfony in the app that runs `schedule:run` might not have the Commands that the worker servers have, they could be different applications entirely 

My personal example is that our cron runs in the frontend server - and the workers server services the message bus. Both are independent symfony apps, and the list of commands on both are different. We would need a way to "mock" the command on the server that injects it into the queue.

**Comment by kbond** [comment]
Have you seen the new symfony/scheduler component? My sort of looong term plan is to deprecate this bundle in favor of it.

**Comment by PhilETaylor** [comment]
yeah I saw it, but I could not find documentation, and with it being less than a month ( Mar 16, 2023) old since creation, it was not really a candidate for "today" haha. 

I'll keep a close eye on it, and will test it once it gets nearer the time. If its as good as your bundle then it will be amazing to have it in core symfony.

**Comment by kbond** [comment]
Yep, very early still. Should be stabilized and documented by 6.4. I won't be deprecating this bundle until 5.4 is EOL. But I also don't think I'll add this feature.
